File: src/epochix/story_engine/phases.py
from __future__ import annotations

from epochix.enums import Phase


def relative_improvement(
    primary_value: float,
    baseline: float,
    *,
    lower_better: bool = False,
    ideal: float | None = None,
) -> float | None:
    """Fraction of the achievable improvement realised so far, in [0, 1].

    Direction-aware: for higher-is-better metrics the ideal defaults to 1.0
    (accuracy/mAP/F1 are normalised to [0, 1]); for lower-is-better metrics
    (loss, MAE, EER, perplexity) the ideal defaults to 0.0 and improvement
    means moving *down* from the baseline. Returns ``None`` when the span is
    degenerate (baseline already at the ideal), so callers can fall back.
    """
    if ideal is None:
        ideal = 0.0 if lower_better else 1.0
    span = (baseline - ideal) if lower_better else (ideal - baseline)
    if abs(span) <= 1e-9:
        return None
    improved = (baseline - primary_value) if lower_better else (primary_value - baseline)
    return max(0.0, min(1.0, improved / span))


def compute_phase(
    progress: float | None,
    primary_value: float,
    baseline: float,
    *,
    lower_better: bool = False,
    ideal: float | None = None,
) -> Phase:
    """Hybrid phase detector using both progress and relative metric improvement.

    Args:
        progress:      0.0–1.0 fraction of training completed, or ``None`` when
                       the total length is unknown (then advancement is driven
                       purely by relative metric improvement).
        primary_value: Current primary metric value (accuracy, mAP, loss, …).
        baseline:      Value at the first recorded epoch.
        lower_better:  True for loss-like metrics where smaller is better.
        ideal:         Optional override for the metric's ideal value.
    """
    rel = relative_improvement(
        primary_value,
        baseline,
        lower_better=lower_better,
        ideal=ideal,
    )
    # When the metric span is degenerate, lean on the clock (or 0 if unknown).
    relative = rel if rel is not None else (progress if progress is not None else 0.0)
    # Advancement = the clock when we have one, else metric-driven progress.
    adv = progress if progress is not None else relative

    if adv < 0.10:
        return Phase.AWAKENING
    if adv < 0.40 or relative < 0.40:
        return Phase.LEARNING
    if adv < 0.70 or relative < 0.75:
        return Phase.UNDERSTANDING
    if adv < 0.95 or relative < 0.95:
        return Phase.MASTERING
    return Phase.POLISHING


def estimate_progress(
    current_epoch: float | None,
    total_epochs: int | None,
    step: int | None = None,
    total_steps: int | None = None,
) -> float | None:
    """Best-effort 0–1 progress fraction, or ``None`` when it can't be known.

    Returning ``None`` (rather than a fabricated constant) lets the phase
    detector advance on real metric improvement instead of pretending the run
    is stuck at the start when no total length was reported.
    """
    if current_epoch is not None and total_epochs and total_epochs > 0:
        return min(current_epoch / total_epochs, 1.0)
    if step is not None and total_steps and total_steps > 0:
        return min(step / total_steps, 1.0)
    return None

File: epochix-vscode/src/story/phases.ts
/**
 * TypeScript port of src/epochix/story_engine/phases.py
 */

export type Phase =
  | "awakening"
  | "learning"
  | "understanding"
  | "mastering"
  | "polishing";

export const PHASE_EMOJIS: Record<Phase, string> = {
  awakening: "🌱",
  learning: "📚",
  understanding: "💡",
  mastering: "🎯",
  polishing: "✨",
};

/**
 * Hybrid phase detector using both progress and relative metric improvement.
 *
 * @param progress      0.0–1.0 fraction of total epochs completed.
 * @param primaryValue  Current primary metric value.
 * @param baseline      Value at epoch 0 (or first recorded epoch).
 * @param target        The ideal value: 1.0 for accuracy/mAP, 0 for a loss.
 * @param lowerBetter   Whether improvement means moving DOWN towards target.
 *                      Without it a falling loss measured towards 1.0 never
 *                      counted as progress at all.
 */
export function computePhase(
  progress: number,
  primaryValue: number,
  baseline: number,
  target: number,
  lowerBetter = false,
): Phase {
  if (progress < 0.10) return "awakening";

  const span = lowerBetter ? baseline - target : target - baseline;
  const moved = lowerBetter ? baseline - primaryValue : primaryValue - baseline;
  const relative = span > 0 ? moved / (span + 1e-9) : progress;

  if (progress < 0.40 || relative < 0.40) return "learning";
  if (progress < 0.70 || relative < 0.75) return "understanding";
  if (progress < 0.95 || relative < 0.95) return "mastering";
  return "polishing";
}

/**
 * Fraction of the achievable improvement realised so far, clamped to [0, 1].
 * Mirrors relative_improvement in story_engine/phases.py; undefined when the
 * baseline already sits at the ideal, so callers can fall back.
 */
export function relativeImprovement(
  primaryValue: number,
  baseline: number,
  lowerBetter: boolean,
): number | undefined {
  const ideal = lowerBetter ? 0 : 1;
  const span = lowerBetter ? baseline - ideal : ideal - baseline;
  if (Math.abs(span) <= 1e-9) return undefined;
  const improved = lowerBetter ? baseline - primaryValue : primaryValue - baseline;
  return Math.max(0, Math.min(1, improved / span));
}

/**
 * Best-effort 0–1 progress estimate from whatever info is available.
 */
export function estimateProgress(
  currentEpoch: number | null,
  totalEpochs: number | null,
  step: number | null = null,
  totalSteps: number | null = null,
): number {
  if (currentEpoch !== null && totalEpochs !== null && totalEpochs > 0) {
    return Math.min(currentEpoch / totalEpochs, 1.0);
  }
  if (step !== null && totalSteps !== null && totalSteps > 0) {
    return Math.min(step / totalSteps, 1.0);
  }
  return 0.05; // non-zero so AWAKENING fires immediately
}
